### PR TITLE
Add tests and various bugfixes

### PR DESCRIPTION
**Pull request by cuminandpaprika**
# Summary
This PR adds some basic unit tests and includes several bug fixes. Currently, the main sysl repo can't import the sysl-catalog master branch due to issues with the BlackFriday dependency.

## Proposed Changes:

- Add basic integration tests which use snapshots to verify generated output
- Add unit tests for query param, path param and return types
- Fix bugs in query param, path param and return type diagram generation

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -15,11 +16,12 @@ require (
 	github.com/joshcarp/mermaid-go v0.0.0-20200403052241-d6fecd5216de
 	github.com/pkg/errors v0.9.1
 	github.com/radovskyb/watcher v1.0.7
-	github.com/russross/blackfriday v2.0.0+incompatible
+	github.com/russross/blackfriday/v2 v2.0.1
```
Comment (joshcarp) [inline go.mod]: v2 doesn't work with go mods: https://github.com/russross/blackfriday/issues/500
Already talked with @cuminandpaprika about this

File: go.mod [hunk 2]
```diff
@@ -15,11 +16,12 @@ require (
...
 	github.com/sirupsen/logrus v1.6.0
 	github.com/spf13/afero v1.2.2
+	github.com/stretchr/testify v1.5.1
 	golang.org/x/net v0.0.0-20200506145744-7e3656a0809f // indirect
 	golang.org/x/sys v0.0.0-20200511232937-7e40ca221e25 // indirect
 	gopkg.in/alecthomas/kingpin.v2 v2.2.6
+	gotest.tools v2.2.0+incompatible
 )
-replace github.com/russross/blackfriday => github.com/russross/blackfriday v1.5.2
```
Comment (joshcarp) [inline go.mod]: see above

File: pkg/catalog/diagram.go [hunk 3]
```diff
@@ -53,7 +52,7 @@ func (p *Generator) CreateMarkdown(t *template.Template, outputFileName string,
 	}
 	out := buf.Bytes()
 	if p.Format == "html" && !p.DisableCss {
-		raw := string(blackfriday.MarkdownCommon(out))
+		raw := string(blackfriday.Run(out))
```
Comment (joshcarp) [inline pkg/catalog/diagram.go]: See first comment

File: pkg/catalog/diagram.go [hunk 4]
```diff
@@ -253,20 +288,12 @@ func (p *Generator) CreateFile(contents string, diagramType int, elems ...string
...
 		return ""
 	}
-	newFileName := fileName
-	for i := 0; ; i++ {
-		if diagram, ok := targetMap[newFileName]; !ok || diagram == fileContents {
-			break
-		}
-		newFileName = strings.ReplaceAll(fileName, p.Ext, strconv.Itoa(i)+p.Ext)
-	}
-	fileName = newFileName
```
Comment (joshcarp) [inline pkg/catalog/diagram.go]: why did you remove this??

File: pkg/catalog/diagram.go [hunk 5]
```diff
@@ -228,14 +265,12 @@ func (p *Generator) CreateTypeDiagram(app *sysl.Application, typeName string, t
 func CreateFileName(dir string, elems ...string) (string, string) {
 	absolutefileName := path.Join(append([]string{dir}, Map(elems, SanitiseOutputName)...)...)
 	relativefileName := strings.Replace(absolutefileName, dir+"/", "", 1)
-	absolutefileName = strings.ToLower(absolutefileName)
-	relativefileName = strings.ToLower(relativefileName)
```
Comment (joshcarp) [inline pkg/catalog/diagram.go]: strings.ToLower is needed so that there aren't filename clashes

File: pkg/catalog/diagram.go [hunk 6]
```diff
@@ -279,48 +306,6 @@ func (p *Generator) GenerateDataModel(app *sysl.Application) string {
 	return p.CreateFile(plantumlString, plantuml, appName, "types"+p.Ext)
 }
 
-// CreateQueryParamDataModel returns a Query Parameter data model filename.
-func (p *Generator) CreateQueryParamDataModel(CurrentAppName string, param *sysl.Endpoint_RestParams_QueryParam) string {
```
Comment (joshcarp) [inline pkg/catalog/diagram.go:302]: ??

File: pkg/catalog/template.go [hunk 7]
```diff
@@ -46,7 +46,7 @@ const NewPackageTemplate = `
 ![]({{CreateSequenceDiagram $appName $e}})
 
 ### Request types
-{{if eq (len $e.Param) 0}}
+{{if and (eq (len $e.Param) 0) (eq $e.RestParams nil) }}
```
Comment (joshcarp) [inline pkg/catalog/template.go]: is this tested?
in go templating you can't equate a pointer type to nil: https://goplay.space/#QOWlt3VKeQr

`nil` is casted to `false`, same as any other zero type

File: pkg/catalog/util.go [hunk 8]
```diff
@@ -135,18 +135,10 @@ func ModulePackageName(m *sysl.Module) string {
...
 	for i, v := range vs {
-		for j, f := range funcs {
-			var middle string
-			if j == 0 {
-				middle = f(v)
-				vsm[i] = middle
-			}
-			vsm[i] = f(middle)
-		}
-
```
Comment (joshcarp) [inline pkg/catalog/util.go]: ??

File: pkg/catalog/diagram.go [hunk 9]
```diff
@@ -253,20 +288,12 @@ func (p *Generator) CreateFile(contents string, diagramType int, elems ...string
...
-		}
-		newFileName = strings.ReplaceAll(fileName, p.Ext, strconv.Itoa(i)+p.Ext)
-	}
-	fileName = newFileName
 	// if p.ImageTags: return image tag from plantUML service
 	if p.ImageTags && diagramType == plantuml {
 		return fileContents
 	}
 	targetMap[fileName] = fileContents
-	return strings.Replace(fileName, p.CurrentDir+"/", "", 1)
```
Comment (joshcarp) [inline pkg/catalog/diagram.go]: this is still needed

**Review by joshcarp — CHANGES_REQUESTED** [review]
I guess it's about time that this repo had some tests, but:
- You've reverted a commit (duplicate files) and removed a whole bunch of functionality that is needed. 
- You've added a bug into the templating (pointer equality to nil)
- The v2 change to blackfriday was very intentional, some users were experiencing problems with using `go get` with v2 (it's well documented on their repo)
- You've added a .snapshot folder with svgs, but plantuml svgs aren't deterministic, so diffs will always occur (is this expected, or are you only planning on checking that the files exist?, and if so, why not just have a set of filenames in the code?)

**Comment by cuminandpaprika** [comment]
Thanks for the feedback, 

- yup, will fix up the reverted change
- ah, I didn’t know they were no deterministic, might need to use string finding instead. Do you know if what parts change in plantuml diagrams? Was it the ordering or the layout perhaps? I vaguely recall that happening with integration diagrams... 
- will fix the nil pointer issue, the error log would be super handy if you have it

Should I setup CI on this repo or are we intending to merge it into sysl in the future?

**Comment by cuminandpaprika** [comment]
Also, is it feasible to replace blackfriday with goldmark instead so we can bypass all these go module issues? https://github.com/yuin/goldmark

It’s what hugo moved to from blackfriday

**Comment by joshcarp** [comment]
- https://goplay.space/#QOWlt3VKeQr is the nil pointer issue; you don't need logs
- plantuml svgs are non deterministic, the plantuml code is (that's actually how sysl-catalog avoids geenerating duplicate diagrams; it stores everything in a map and if there's a match for the content it returns that filename)
- Which markdown generator is used doesn't matter

File: go.mod (hunk 2 above)
Comment (cuminandpaprika) [reply go.mod]: Removed

File: go.mod (hunk 1 above)
Comment (cuminandpaprika) [reply go.mod]: 👍 Removed

File: pkg/catalog/diagram.go (hunk 3 above)
Comment (cuminandpaprika) [reply pkg/catalog/diagram.go]: 👍 Removed

File: pkg/catalog/diagram.go (hunk 4 above)
Comment (cuminandpaprika) [reply pkg/catalog/diagram.go]: 👍 Dropped that commit

File: pkg/catalog/diagram.go (hunk 5 above)
Comment (cuminandpaprika) [reply pkg/catalog/diagram.go]: 👍 Dropped that commit

File: pkg/catalog/diagram.go (hunk 6 above)
Comment (cuminandpaprika) [reply pkg/catalog/diagram.go:302]: Bungled rebase I think, fixed up in latest commits

File: pkg/catalog/template.go (hunk 7 above)
Comment (cuminandpaprika) [reply pkg/catalog/template.go]: Great catch, changed it to `with` instead of `if`

File: pkg/catalog/util.go (hunk 8 above)
Comment (cuminandpaprika) [reply pkg/catalog/util.go]: Dropped this commit

File: pkg/catalog/diagram.go (hunk 9 above)
Comment (cuminandpaprika) [reply pkg/catalog/diagram.go]: Dropped this

**Comment by joshcarp** [comment]
```
app:
    ep(req <: app2.this):
        return ok <: app2.this
app2:
    !type this:
        contents <: string
```

sysl-catalog --serve example.sysl:
image

**Comment by cuminandpaprika** [comment]
> ```
> app:
>     ep(req <: app2.this):
>         return ok <: app2.this
> app2:
>     !type this:
>         contents <: string
> ```
> 
> sysl-catalog --serve example.sysl:
> image

Thanks for finding that, fixed up along with a bunch of other issues with diagram generation which that helped me uncover!

**Review by joshcarp — APPROVED** [review]
